### PR TITLE
Add trophy with tooltip to championships

### PR DESCRIPTION
**Pull request by UnsolvedCypher**
This is a potential fix for #1983. A potential issue with this is that it isn't super obvious that the trophy is hoverable, but I think this is a problem that much of the WCA website faces. A possible addition might be a little badge to the right of the competition name that says what kind of championship it is.

screenshot from 2018-12-19 22-54-52

### CODE REVIEW

File: WcaOnRails/app/assets/stylesheets/competitions.scss [hunk 1]
```diff
@@ -397,3 +397,7 @@ table.admin-competitions {
     }
   }
 }
+
+.championship-trophy {
+  color: #e64503;
```
Comment (jonatanklosko) [inline WcaOnRails/app/assets/stylesheets/competitions.scss]: Perhaps this should be either of https://github.com/thewca/worldcubeassociation.org/blob/793e9cd4a2e391a41593b7bb5761e43395049500/WcaOnRails/app/assets/stylesheets/variables.scss#L9-L10
If not, then feel free to add a new variable instead of hard-coding the color.

File: WcaOnRails/app/views/competitions/_nav.html.erb [hunk 2]
```diff
@@ -169,7 +169,14 @@
 
     </div>
     <div id="competition-data">
-      <h3><%= @competition.name %></h3>
+      <h3>
+        <% if @competition.championships.any? %>
+          <span class="championship-trophy" data-toggle="tooltip" data-placement="bottom" title="<%= @competition.championships.map { |championship| championship.name }.join(", ") %>">
```
Comment (jonatanklosko) [inline WcaOnRails/app/views/competitions/_nav.html.erb]: `Championship#name` is not internationalized yet, if we want to display that on competition pages, we definitely want to do so now.

Comment (UnsolvedCypher) [reply WcaOnRails/app/views/competitions/_nav.html.erb]: Good call. I'm having trouble understanding how the country names are internationalized, since they aren't in the translation yamls. Do I need to query the database to get those?

File: WcaOnRails/app/assets/stylesheets/competitions.scss (hunk 1 above)
Comment (UnsolvedCypher) [reply WcaOnRails/app/assets/stylesheets/competitions.scss]: I was going for the WCA orange that's used throughout the site, which I found was $link-color. I'll make that change.

File: WcaOnRails/app/views/competitions/_nav.html.erb (hunk 2 above)
Comment (jonatanklosko) [reply WcaOnRails/app/views/competitions/_nav.html.erb]: We have this gem `i18n-country-translations` which I believe handles most names, and some additional keys in the `en.yml` file: https://github.com/thewca/worldcubeassociation.org/blob/793e9cd4a2e391a41593b7bb5761e43395049500/WcaOnRails/config/locales/en.yml#L24-L31
The `Country` model includes `LocalizedSortable` which in turn gives it this method: https://github.com/thewca/worldcubeassociation.org/blob/793e9cd4a2e391a41593b7bb5761e43395049500/WcaOnRails/app/models/concerns/localized_sortable.rb#L10-L12
So basically `Country#name` is internationalized, but there's also `Country#name_in` if you want to force some language (e.g. just English for non-i18n parts).

Comment (UnsolvedCypher) [reply WcaOnRails/app/views/competitions/_nav.html.erb]: That's very helpful, thank you. So it looks like continent.name actually already calls the method you're talking about, but I think unfortunately those translations aren't always going to work since they're in the nominative case. Do we already have some way of dealing with this situation? For example, in Russian you might say "Чемпионат Индии" (genitive case), while the translation for "India" is "Индия", which is in the nominative case.

Comment (jfly) [reply WcaOnRails/app/views/competitions/_nav.html.erb]: Yeah, "national championship: [country]" or "National championship of [country]" sounds good to me!

Comment (UnsolvedCypher) [reply WcaOnRails/app/views/competitions/_nav.html.erb]: "National Championship for [country]" works great in English and some other languages, but it doesn't work in languages with a case system (such as Russian or German). In languages like this, the noun is modified (in part) according to the preposition, so "[country]" and "of [country]" require different translations of the country. This is a rough analogy, but it would sound something like "National Championship for French" for example.

"National championship: [country]" is a little awkward, but could work. The alternative is adding a bunch of translation keys for the countries in the genitive case, but that is quite a bit of work for translators.

Comment (jfly) [reply WcaOnRails/app/views/competitions/_nav.html.erb]: Ah, ok! How about we do "National championship: [country]" for now, and we can decide later if we want to make our translators do a bunch of these translations to make it less awkward sounding.

**Comment by EdHollingdale** [comment]
How would this work for e.g. WC2013 which was the World Championship and the National Championship for USA?

**Comment by EdHollingdale** [comment]
I feel that there should be an order of World > Continental > National (I'd say something like greater China would come under National). It'd be good if this were implemented with the trophies as it will likely be required for WC2019.

**Comment by UnsolvedCypher** [comment]
I would be happy to take care of the order- is there a preferred order for this?

**Comment by jfly** [comment]
Your call. If you don't have an opinion, then the order @EdHollingdale suggests seems fine to me.

**Comment by EdHollingdale** [comment]
World > Continental > National is consistent with how it is done on the championship podiums section in profiles.

**Comment by UnsolvedCypher** [comment]
Sorry for the delay on this. This is almost done, but I'm getting "comparison of Championship with Championship failed" after making Championships comparable and sorting them before displaying. Is making them comparable the right approach here, or is that overkill for the task at hand? I will try to do some more investigating of this issue, but I assume the problem is that my method is implemented wrong.

**Comment by UnsolvedCypher** [comment]
Ok, sorry that took a few commits- that tied into a few places in the tests where I didn't expect it. It should be working now. I also added a few methods to `Championship` to make it easier to determine what kind it is. This cleans up the comparator a lot, and also may be useful in the future. If you think this is excessive, I can make the variables local. Also, I'm not sure if the i18n keys are in the best place, but I couldn't find anywhere better for them to go.

**Comment by EdHollingdale** [comment]
Do you know what the timeline is for this appearing on the site? It would be great to have this implemented for WC2019 announcement...

**Comment by UnsolvedCypher** [comment]
@EdHollingdale this is ready to be merged in as soon as a WST member has time to look at it. If this is time-sensitive, you may want to email the WST and let them know.

File: WcaOnRails/app/models/championship.rb [hunk 3]
```diff
@@ -15,16 +16,46 @@ class Championship < ApplicationRecord
...
     country = Country.find_by_iso2(championship_type)
-    return "National Championship for #{country.name}" if country
+    return I18n.t('competitions.competition_form.championship_types.national', country: country.name) if country
 
-    "Championship for #{championship_type}"
+    I18n.t('competitions.competition_form.championship_types.generic', type: championship_type)
+  end
+
+  def national?
+    Country.find_by_iso2(championship_type)
```
Comment (jfly) [inline WcaOnRails/app/models/championship.rb]: nit: These helper methods have some duplicated code from `def name` up above. It *might* be nice to refactor things so `name` can use these as well. I'm thinking something like: `def country`, which returns the country if it's a national championship, and nil if it's not.

Feel free to agree or disagree with me here about if this makes the code better.

File: WcaOnRails/app/models/championship.rb [hunk 4]
```diff
@@ -15,16 +16,46 @@ class Championship < ApplicationRecord
...
+
+  def continental?
+    Continent.c_find(championship_type)
+  end
+
+  def world?
+    championship_type == "world"
+  end
+
+  def <=>(other)
```
Comment (jfly) [inline WcaOnRails/app/models/championship.rb]: I always find these sorts of methods impossible to read when written this way. A pattern I've been very fond of is to munge the object into an array, which you can then compare lexicographically. See https://stackoverflow.com/a/28558053/1739415 for an example of this. In this case, I think we'd do something like this:

```ruby
def to_a
  [
    world? ? 0 : 1,
    continental? ? 0 : 1,
    greater_china? ? 0 : 1,
    national? ? 0 : 1,
  ]
end

def <=>(other)
  self.to_a <=> other.to_a
end
```

This pattern may feel a bit weird at first, but I like it because it's both explicit and concise about what attributes take precedence over what other things.

Comment (UnsolvedCypher) [reply WcaOnRails/app/models/championship.rb]: Wow, this is super legible. I don't think I've ever seen a particularly legible `compare_to` function that wasn't trivial. I'll definitely do this, thanks!

File: WcaOnRails/app/models/championship.rb (hunk 3 above)
Comment (UnsolvedCypher) [reply WcaOnRails/app/models/championship.rb]: Yeah, this did cross my mind,  but that leaves me with the question of how to handle world and Greater China. For word I suppose I could do something like `def planetary` but that seems a bit goofy given that I don't think we anticipate any other planetary championships. I really have no idea what to do for Greater China. Or, would you suggest just leaving those two as booleans and just having country and continent?

Comment (jfly) [reply WcaOnRails/app/models/championship.rb]: I think for Greater China, the `greater_china?` method you've already written fits in fine here. Just change the existing line of code to:

```ruby
return I18n.t('competitions.competition_form.championship_types.greater_china') if greater_china?
```

If we ever add a second type of cross-country championship type, we can revisit this and maybe do something like:

```ruby
return I18n.t("competitions.competition_form.championship_types.#{championship_type}") if cross_country_championship?
```

but this might require a bit of i18n work to avoid confusing our scripts that detect if all keys are defined and used, so I don't think it's worth doing right now.

Comment (jfly) [reply WcaOnRails/app/models/championship.rb]: For worlds, I would suggest doing something similar, just rewrite the code to do:

```ruby
return I18n.t('competitions.competition_form.championship_types.world') if world?
```

Comment (UnsolvedCypher) [reply WcaOnRails/app/models/championship.rb]: Ok, so to clarify, you are suggesting that I leave the `world?` and `greater_china?` boolean methods and also add the string-returning methods here? And same for country/continent? Or were you just using `... if world?` as shorthand and that would be replaced by the code checking to see if it's a world championship?

Comment (jfly) [reply WcaOnRails/app/models/championship.rb]: Oh, sorry, I was suggesting real code that could replace the existing code in the `name` method up above. No changes to the `world?` and `greater_china?` method you've already written.

**Comment by UnsolvedCypher** [comment]
@jfly just pinging you to let you know that the checks have passed and this is ready (since I know the Worlds team wants this soon).

**Comment by jfly** [comment]
@UnsolvedCypher, please see my message on slack.

**Comment by jfly** [comment]
Great work, thanks @UnsolvedCypher!

**Comment by EdHollingdale** [comment]
This looks great! Great job @UnsolvedCypher and @jfly
